cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,548 @@
+    if (srcReplicaToSwap.isLeader() && dstReplicaToSwap.isLeader() && Objects.equals(srcTopicName, dstTopicName)) {
+      // Swapping two leader replicas of the same topic is always accepted
+      return ACCEPT;
+    }
+    if (doesLeaderReplicaRemoveViolateOptimizedGoal(srcReplicaToSwap) || doesLeaderReplicaRemoveViolateOptimizedGoal(dstReplicaToSwap)) {
+      return REPLICA_REJECT;
+    }
+    return ACCEPT;
+  }
+
+  /**
+   * Check whether this goal is violated after removing the given leader replica. It is violated if there is not be enough
+   * leader replicas on that broker for that topic
+   *
+   * @param replicaToBeRemoved replica to be removed
+   * @return {@code true} if the given replica move would violate the optimized goal (i.e. the move is not acceptable),
+   * {@code false} otherwise.
+   */
+  private boolean doesLeaderReplicaRemoveViolateOptimizedGoal(Replica replicaToBeRemoved) {
+    if (!replicaToBeRemoved.isLeader()) {
+      // Moving a follower replica does not violate/affect this goal
+      return false;
+    }
+    String topicName = replicaToBeRemoved.topicPartition().topic();
+    if (!_mustHaveLeaderReplicaPerBrokerTopics.contains(topicName)) {
+      // Moving a replica of a irrelevant topic does not violate/affect this goal
+      return false;
+    }
+    int topicLeaderReplicaCountOnSourceBroker = replicaToBeRemoved.broker().numLeaderReplicasOfTopicInBroker(topicName);
+    return topicLeaderReplicaCountOnSourceBroker <= minTopicLeadersPerBroker();
+  }
+
+  /**
+   * This is a hard goal; hence, the proposals are not limited to dead broker replicas in case of self-healing.
+   * Sanity Check: The total number of leader replicas of certain topics is sufficient so that these leader replicas
+   * can be moved to satisfy this goal.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    _mustHaveLeaderReplicaPerBrokerTopics = Collections.unmodifiableSet(
+        Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel::topics));
+    if (_mustHaveLeaderReplicaPerBrokerTopics.isEmpty()) {
+      return;
+    }
+    // Sanity checks
+    validateTopicsWithMinLeaderReplicaIsNotExcluded(clusterModel, optimizationOptions);
+    validateEnoughLeaderReplicaToDistribute(clusterModel, optimizationOptions);
+    validateBrokersAllowedReplicaMoveExist(clusterModel, optimizationOptions);
+    // Sort leader replicas for each broker based on resource utilization.
+    new SortedReplicasHelper().addPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants())
+                              .trackSortedReplicasFor(replicaSortName(this, false, false), clusterModel);
```

Comment:
Comment and the sorted replicas do not match -- i.e. the sorted replicas is incorrect (please see other goals for example use).
For example, `optimizationOptions.onlyMoveImmigrantReplicas()` must be honored if it is `true`.

Reply:
Thank you for pointing it out. I changed it to the below. Let me know if it is correct this time:
```java
    boolean onlyMoveImmigrantReplicas = optimizationOptions.onlyMoveImmigrantReplicas();
    Set<String> excludedTopics = optimizationOptions.excludedTopics();
    new SortedReplicasHelper().addSelectionFunc(ReplicaSortFunctionFactory.selectLeaders())
                              .maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants(), onlyMoveImmigrantReplicas)
                              .addSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnExcludedTopics(excludedTopics))
                              .maybeAddPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants(), !onlyMoveImmigrantReplicas)
                              .trackSortedReplicasFor(replicaSortName(this, true, true), clusterModel);
```

Reply:
```
.addSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnExcludedTopics(excludedTopics))
```

Should we add a new selection function to choose from replicas of relevant topics?

Reply:
Good idea